### PR TITLE
refactor and cleanup cilium helm chart

### PR DESCRIPTION
**Pull request by dungdm93**
Signed-off-by: Đặng Minh Dũng <dungdm93@live.com>

Cilium helm charts is quite large (>70 files) with lots of components but all that file are in flatten structure. Some how, it make hard to organized and hard to add new feature. So this PR aim to refactor and cleanup cilium helm chart with following notable changes:
1. **Restructure folder**: move each component in it own folder
Screenshot from 2021-07-02 10-22-23
2. **Remove deprecated & unused resources**:
  * Delete `hubble-ca-configmap.yaml`, it has bean marked as deprecated and will be removed in v1.11
  * Delete `cilium-etcd-sa` Service Account, `etcd-operator` ClusterRole/ClusterRoleBinding which are use in nowhere.
3. **Refactor tls logic**: provided, helm and cronjob.  
  Currently, this chart support 3 different kind of TLS: provided, helm-gen, cronjob-gen. But the logic between those 3 types is not the same, for example: helm-gen method always generate both CA and certs (ignore user provided), but cronjob-gen reuse and disable generate cert when user provide CA/cert (this mean no cert auto-rotate).
https://github.com/cilium/cilium/blob/703b38fd47da1f91a342a8546f53542d624d8b22/install/kubernetes/cilium/templates/_hubble-generate-certs-job-spec.tpl#L47-L53  
  So I simplify and unify this logic as following:
    * Split each tls method to its own folder:
    Screenshot from 2021-07-02 10-45-30
    * When tls provided (`not auto.enabled`), both `cert` and `key` are required
    * When tls auto (both helm and cronjob methods), if user provide CA, use it; else generate new one. Other client/server/remote TLS are generated.
4. cleanup code
    * Move `nodeinit` prestop and startup scripts to external file (inside `files/nodeinit` folder) aim to better highlight and easier to update
    * Indent go template and use `nindent` (instead of `indent`) for better look and easier to collapse/expand when using IDE
    * Using `include "cilium.image"` named template for generate docker image name
    * Support `priorityClassName`. `priorityClassName` is defined in `values.yaml` but used in nowhere. Now it's supported.

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Comment by gandro** [comment]
This is amazing work, and I think I agree with all improvements. However, due to the large size of the PR and the fact that we have little static safety checks on Helm, this will take a while to be reviewed. I therefore a have a request:

Could you split you this PR into a few smaller commits? For example have one commit per newly added folder (i.e. one commit for all `cilium-agent/` related changes, one commit for all `hubble/` related changes etc)? I think this will make the review such much easier.

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:28]: I haven't a real review yet, but this caught my eye - please do not remove these. They are here on purpose.

All `.Capabilities.KubeVersion` checks must be overwritable via `.Values.Capabilities.KubeVersion`, see https://github.com/cilium/cilium/pull/14778

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:100]: As mentioned above, we do want to use `$k8sVersion` here, not `.Capabilities.KubeVersion.Version`

**Review by gandro — CHANGES_REQUESTED** [review]

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml
Comment (dungdm93) [reply install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:28]: Hello @gandro 

`.Values.Capabilities.KubeVersion` was used to generate `quick-install.yaml`, but since cilium 1.10 `MIN_K8S_VERSION` is not using in `install/kubernetes/Makefile` anymore.

Furthermore, since 3.6, `helm template` support set Kubernetes version used for Capabilities.KubeVersion by using [`--kube-version string`](https://helm.sh/docs/helm/helm_template/). Do you think it's safe to remove this workaround?

/cc @aanm

**Comment by dungdm93** [comment]
> Could you split you this PR into a few smaller commits? For example have one commit per newly added folder (i.e. one commit for all cilium-agent/ related changes, one commit for all hubble/ related changes etc)? I think this will make the review such much easier.

Thanks for your suggestion. I going to close this one and split it into few smaller PR

**Comment by gandro** [comment]
> > Could you split you this PR into a few smaller commits? For example have one commit per newly added folder (i.e. one commit for all cilium-agent/ related changes, one commit for all hubble/ related changes etc)? I think this will make the review such much easier.
> 
> @gandro Thanks for your suggestion. I going to close this one and split it into few smaller PR

Ah! So I was actually suggesting to have a single PR with multiple commits. But multiple PRs also work, and might be even better.

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml
Comment (gandro) [reply install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:28]: That's a fair point. I think we have a few users using `helm template`, so it's not just the quick-install.yaml - but I wasn't aware of the feature in Helm 3.6, that will be a good replacement going forward.

But I'll let @aanm chime in here as well. Removing Helm values is usually a breaking change, so we if we remove this, we want to communicate this clearly.

Comment (aanm) [reply install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:28]: @dungdm93 @gandro For reference, the commit that introduced this change was https://github.com/cilium/cilium/commit/4cc699dd8d99877b8456c9246ab64a741bee7030. So yeah it looks that it is fine to remove these lines as long we update the requirement of helm to >= 3.6.

Comment (dungdm93) [reply install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:28]: Thank @aanm, I will make this change right after #16795 is merged

**Pull request closed without merging**
